OpenRA.Mods.Common/Activities/PickupUnit.cs

```diff
@@ -74,24 +74,18 @@ public override bool Tick(Actor self)
 				return true;
 			}
 
-			if (carryall.State != Carryall.CarryallState.Reserved)
-				return true;
-
-			switch (state)
+			var distance = (cargo.CenterPosition - self.CenterPosition).HorizontalLength;
+			if (distance <= 4096 && !locked)
 			{
-				case PickupState.Intercept:
-					QueueChild(movement.MoveWithinRange(Target.FromActor(cargo), WDist.FromCells(4)));
-					state = PickupState.LockCarryable;
-					return false;
-
-				case PickupState.LockCarryable:
-					if (!carryable.LockForPickup(cargo, self))
-						Cancel(self);
-
-					state = PickupState.Pickup;
-					return false;
-
-				case PickupState.Pickup:
+				var lockResponse = carryable.LockForPickup(cargo, self);
+				if (lockResponse == LockResponse.Failed)
+					Cancel(self);
+				else if (lockResponse == LockResponse.Pending && distance <= 1024)
```

Comment:
This will break as soon as the `FlyIdle` makes the carryall move more than 1 cell away. See my linked commit for a workaround.

Reply:
The `FlyIdle` could in principle move it even outside the initial 4 cell trigger, breaking things even further. We may need to go back to explicit stages for moving to target, idling, and pickup.